backend/security/src/main/kotlin/io/tolgee/security/authentication/AuthenticationFilter.kt

```diff
@@ -0,0 +1,131 @@
+/**
+ * Copyright (C) 2023 Tolgee s.r.o. and contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.tolgee.security.authentication
+
+import io.tolgee.component.CurrentDateProvider
+import io.tolgee.configuration.tolgee.AuthenticationProperties
+import io.tolgee.constants.Message
+import io.tolgee.exceptions.AuthenticationException
+import io.tolgee.security.PAT_PREFIX
+import io.tolgee.security.ratelimit.RateLimitService
+import io.tolgee.service.security.ApiKeyService
+import io.tolgee.service.security.PatService
+import io.tolgee.service.security.UserAccountService
+import org.springframework.security.core.context.SecurityContextHolder
+import org.springframework.stereotype.Component
+import org.springframework.web.filter.OncePerRequestFilter
+import javax.servlet.FilterChain
+import javax.servlet.http.HttpServletRequest
+import javax.servlet.http.HttpServletResponse
+
+@Component
+class AuthenticationFilter(
+  private val authenticationProperties: AuthenticationProperties,
+  private val currentDateProvider: CurrentDateProvider,
+  private val rateLimitService: RateLimitService,
+  private val jwtService: JwtService,
+  private val userAccountService: UserAccountService,
+  private val apiKeyService: ApiKeyService,
+  private val patService: PatService,
+) : OncePerRequestFilter() {
+  override fun doFilterInternal(request: HttpServletRequest, response: HttpServletResponse, filterChain: FilterChain) {
+    val policy = rateLimitService.getIpAuthRateLimitPolicy(request)
+
+    if (policy == null) {
+      doAuthenticate(request)
+    } else {
+      rateLimitService.consumeBucketUnless(policy) {
+        doAuthenticate(request)
```

Comment:
No because it's an inverted check. If the lambda returns true, no token is consumed. Hence the unless name.

It was done this way because throwing an exception does abort processing and the mental model implies to still consume a bucket because it didn't return true.

This is unit tested.